tensorboard/plugins/histogram/tf_histogram_dashboard/histogramCore.ts

```diff
@@ -86,10 +86,22 @@ export function intermediateToD3(
   max: number,
   numBins = 30
 ): D3HistogramBin[] {
+  // Empty case.
+  if (!histogram.buckets || histogram.buckets.length === 0) {
+    return [];
+  }
+  // Single value case.
```

Comment:
So, but while this does render something, which is better than nothing at all... it's a little hard to interpret (especially in the offset view, where it's just kind of a blank white vertical line until you hover over it).

Kind of like the empty buckets case above, it might produce a cleaner result (and more like Time Series) if instead of special-casing the single value case here, we leave in place the existing logic to compute a spaced-out min and max.  (It's ok that this doesn't match the histogram v3 requirements; those are only about the canonical "persisted" form of the data, and at this point in the code, we're optimizing for the visualization itself, so the reformatting is only about presentation.)  Instead, we can modify how the intersection is computed below so that we still count input buckets with a zero width as contributing to the output bins.  This is basically the same as what the newer Time Series logic does to calculate the bin's "contribution".

I haven't fully tested it but here's a sample patch to the body of the while loop to show the suggestion:

```patch
@@ -119,12 +112,20 @@ export function intermediateToD3(
       // infinite-sized.
       const bucketRight = Math.min(max, histogram.buckets[bucketIndex].right);
       const bucketLeft = Math.max(min, histogram.buckets[bucketIndex].left);
-      const intersect =
-        Math.min(bucketRight, binRight) - Math.max(bucketLeft, binLeft);
-      const count =
-        (intersect / (bucketRight - bucketLeft)) *
-        histogram.buckets[bucketIndex].count;
-      binY += intersect > 0 ? count : 0;
+      const bucketWidth = bucketRight - bucketLeft;
+      if (bucketWidth > 0) {
+        const intersect =
+          Math.min(bucketRight, binRight) - Math.max(bucketLeft, binLeft);
+        const count =
+          (intersect / (bucketRight - bucketLeft)) *
+          histogram.buckets[bucketIndex].count;
+        binY += intersect > 0 ? count : 0;
+      } else {
+        const isFinalBin = binRight >= max;
+        const singleValueOverlap = (binLeft <= bucketLeft
+            && (isFinalBin ? bucketRight <= binRight : bucketRight < binRight));
+        binY += singleValueOverlap ? histogram.buckets[bucketIndex].count : 0;
+      }
       // If `bucketRight` is bigger than `binRight`, then this bin is
       // finished and there is data for the next bin, so don't increment
       // `bucketIndex`.```

Reply:
Thanks for the further explanation and clarifying that we should focus on the visualization purpose here!

- For the single data case, `min` would equal `max`, and it will never enter the `d3.range(min, max, binWidth)` loop, replaced `d3.range` with a for loop.
- Also applied an offset to `minmin` and `maxmax` in `backendToVz()` if they are equal like the timeseries approach.

Reply:
Just curious, is there a reason to apply the offsetting to minmin and maxmax in `backendToVz()` instead of here in `intermediateToD3()`, which is where it was before?  It seems simpler to just revert to what's currently implemented, which matches timeseries:

https://github.com/tensorflow/tensorboard/blob/165f2cbd2a44cc74d041223653cc447666e07ec8/tensorboard/plugins/histogram/tf_histogram_dashboard/histogramCore.ts#L89-L93

If we put this logic below the `if (min === undefined || max == undefined) {` block that you added, then we will also handle that case, rather than leaving one edge case in which `min === max` (since they're both remapped from undefined to 0).  I suppose if you prefer, we could move the undefined handling into `backendToVz()` and then re-tighten the arguments here so that they can't be `undefined` after all - but then we still kind of need some check here to confirm that the arguments aren't equal, which is why doing it here makes a little more sense to me.

That way, we can ensure that `intermediateToD3()` is consistently always returning bins that aren't zero-width, which is easier for the visualization to meaningfully interpret.

Reply:
Thanks! For some reason after the last iteration I forgot this `[*|/] 1.1 [+|-] 1` existed here and thought it only existed in the Timeseries one.